### DISCUSSION TITLE
Improve documentation and provide better samples

### DISCUSSION THREAD

**Discussion by empz** [fluentmigrator/fluentmigrator#1398] [Q&A]
I'm trying to integrate fluentmigrator in our solution built of a couple of class libraries and an ASP.NET Core 3.1 app. We're using Dapper with SqlServer.

We have a couple of questions that could n ot find an answer for in the docs or here:

1. Docs says that the in-process is recommended. Does that mean that we should plugin fluentmigrator libraries right into our ASP.NET Core application instead of having a separate Console project that will be responsible for running the migrations?
2. It's not clear to us how to rollback to a previous version using a CI/CD approach when something goes wrong.
3. We couldn't find in the docs how to specify a Profile (i.e. Development) that will run certain data migrations on our local dev. environments.

Can we get better documentation and samples?

**Comment by jzabroski** [comment]
> 1. Does that mean that we should plugin fluentmigrator libraries right into our ASP.NET Core application instead of having a separate Console project that will be responsible for running the migrations?

No.  The short answer is, "It depends."  If you're not the one calling the shots in terms how how you deploy applications, then you might have to put FluentMigrator inside the ASP.NET Core application.  For example, if your organization mandates everything run on Azure Functions, then your options are going to be very limited with how you can deploy FluentMigrator.  Another example to consider is if you can only run Azure SQL cloud database, you can't run FluentMigrator using localhost, because you can't deploy FluentMigrator "locally" to a serverless cloud database.  However, I don't publish guidance on how to change the connection string :)  I understand there are probably some funky cloud architecture decisions people have to live with.  I think, rather than improving the documentation, you should ask questions about your specific requirements, because I've tried hard in various projects to "improve the documentation" only to realize that the customer was pitching me an [xy problem](https://xyproblem.info/).  Once we agree what solution works best for you, ideally _you_ go document it for others as a way of saying thanks.

If you do want to deploy FluentMigrator to, say, a serverless fabric, I'd probably want to use [Azure Durable Task Framework](https://github.com/Azure/durabletask) so that if a server reboots in the middle of a migration, I can just seamlessly continue (if I'm running in Transaction Per Migration mode).  But I'd only push for it on a very large, very high budget project.

That said, the _reason_ we recommend using an In-Process Runner is it avoids a flaw in the .NET Project System and RollForward technology, where plug-ins loaded out-of-process (e.g., FluentMigrator assemblies) can have transitive assembly dependencies with different major version numbers than those the FluentMigrator process (e.g., FluentMigrator.DotNet.Cli) use.  Loading an assembly at runtime with a suitable version number is something Microsoft didn't have great user stories for.  You can side-step this problem by creating an in-process runner, which forces MSBuild and the C# Project System to calculate your deps.json file _with your plug-ins_, because _you're plug-ins are no longer plug-ins but part of the executable_.

The simplest example would be `System.ComponentModel.Annotations` nuget package, which has a different patch number for net48 and netstandard2.0 assemblies for the same nuget package version. Despite the fact netstandard2.0 is supposed to be quasi-compatible with net48, it just takes one patch number difference across the various targets to cause issues.  If you don't use an in-process runner, you either need to use FluentMigrator with zero dependencies, or need to write assembly resolve event handlers to tell the runtime which assembly to load.  That is a _lot less safer_ than just having the .NET Project System do all the work for you, because if it is wrong, the C# compiler will likely catch it when it tries to resolve symbols.

> 2\. It's not clear to us how to rollback to a previous version using a CI/CD approach when something goes wrong.

When you say "using a CI/CD approach", what are you referring to?  The Phoenix Project model where the moment code is merged and passes automated quality gates, it gets pushed to production?  Again, this is usually an organization decision way above the person asking the question, but it really dictates what makes sense.  There is no one-size-fits-all solution.  I have my best practices I've groomed over the years, but that doesn't guarantee they're "the best" or fit your model.

I have never used the Phoenix Project model.  In fact, I actively avoid applying to places where they use it.  My current model is:

1. Release every 2 weeks.
2. CI/CD pipeline where I use Red Gate SQL Clone to clone the production database nightly.
3. Every developer gets a private integration environment with a clone of the production database, refreshed nightly.
4. Unit Tests on all Repository objects, testing Create, Read, Update, Delete, BulkInsert, BulkUpdate and BulkMerge.  I call these unit tests since each test runs inside a transaction and no other test can see the state of that test's database transaction.
5. Three fully operational environments: Alpha (master branch), Beta (release branch), Production (latest release approved by QA)

In terms of _my_ best practices:

1. **Don't use Profiles in FluentMigrator; use Maintenance Migrations instead (they are more general)**
2. **Migration version numbers should follow yyyyMMddZZZZ numbering convention, where ZZZZ is a leading 0 padded four digit number, and if you have multiple teams working in the same FluentMigrator assembly, I recommend assigning `0009`-`0999` to one team, `1000`-`1999` to another team, etc.**
    - Some people dislike this guidance, but I have used it on massive database refactorings for a client with 1,000 tables, and we have written over 3,000 migrations this way.
    - I like my guidance because it gives room for hotfixes and continuing work on the mainline branch.  It also gives you room to split up a single, large Migration into many smaller migrations, so that the transaction log writer has time to "catch up" on commit.  The risk is that somebody could "squeeze in" an earlier migration number and create chaos.  In practice, this has never happened to me even with a team of junior developers.
3. **Separate out "data migrations" from "schema migrations"**
4. **Create Tags for every environment.**
    - I use `UnitTest`, `Development`, `Alpha`, `Beta`, `Production` and occasionally an ad-hoc tag for special production deployments that require a server reboot.
    - Example of special migrations that may require a server reboot or some "destructive" operation to physical database connections would include changing AD Group membership, which requires resetting all active connections to a database (to clear our sys.login_tokens).  I usually run permission changes during a maintenance window, and try to keep permissions relatively simple.
5. **Use Tags for "data migrations", as data migrations don't make any sense to run against Tag Name = `UnitTest`**

Now, in terms of "it's not clear to us how to rollback", as a consultant who puts in place best practices, usually people say this because they expect a database migration tool to be somewhat magical.  FluentMigrator has Up/Down migrations, and an explicit rollback command.  If you want a fully automated rollback process (where the process is _self-healing_ and knows what Migration number to rollback to), then yes, you will need to figure out how to maintain the previous highest migration that ran, and ensure nobody commits a migration number with a lower number than anything that ran previously.  _I'm not going to spend time writing guidance on how to do this, as this is heavily custom stuff people would pay me money for._

In terms of "changes requiring rollback", it is better to think through how to avoid needing to rollback than to actually rollback.  In my experience, if the database migration is at fault, it comes down to a few root causes in SQL Server (this is SQL Server specific):
   - You're not using a tool like RedGate Developer Toolbelt and using the awesome `Find Invalid Objects` feature in your test environment, prior to ever deploying (this is an easy victory you should do regardless of whether you use FluentMigrator)
       - This is useful if you are working on legacy systems that have lots of stored procedures, triggers, and views.  If you're just using tables, you don't have any "modules" (Transact-SQL code stored on the server) to refresh.
   - You're not using sp_refreshsqlmodule to ask SQL Server to refresh transitive schema dependencies, such as view column metadata.  I compute the transitive closure of all dependencies as a directed acyclic graph, and walk the DAG using depth first search, and call sp_refreshsqlmodule on every node.  (This works **way** better on SQL Server 2017 and later, as earlier versions of SQL Server had bugs in sys.sql_dependencies that ApexSQL reported to Microsoft viz a viz an early whitepaper arguing why their T-SQL parser is superior to Microsoft's T-SQL parser that populates sys.sql_dependencies.)
       - This should be a **requirement** in any environment that uses schema replication or HADR.
       - This is useful if you are working on legacy systems that have lots of stored procedures, triggers, and views.  If you're just using tables, you don't have any "modules" (Transact-SQL code stored on the server) to refresh.  If you're not using schema replication and you're only using tables, there is nothing to refresh to synchronize.
   - You're using `SNAPSHOT ISOLATION` because somebody thought it would resolve issues with bad code causing deadlocks.  Don't use `SNAPSHOT ISOLATION` and expect sensible database migrations.  If you have to use it, disable it when upgrading the database schema.  If you read the documentation closely, `SNAPSHOT ISOLATION` changes the transaction isolation level of transactions from `READ COMMITTED` to something called multi-version concurrency control (MVCC).  While there are databases where you get MVCC _and_ read committed transactions (Postgres), putting the database in SNAPSHOT ISOLATION mode effectively stops Transact-SQL from being a Transact-ional SQL DDL.
   - You're using `sp_` undocumented SQL Server functions, and those functions contain bugs.
       - For example, `sp_rename` has different bugs in different versions of SQL Server, so renaming things is risky, especially if you don't keep identical SQL Server versions between your CI/CD environment database server, and your production database server.   One particular example is that `sp_rename` doesn't work under certain conditions when the object being renamed is associated with a temporal table.
       - For example, `sp_addrolemember` and `sp_droprolemember` rely on `sys.sp_validname`, which can contain bugs.  Ideally, you use `ALTER ROLE roleName DROP MEMBER databasePrincipalName`.
   - You're using FluentMigrator to upgrade a database to a newer version of SQL Server, but not using Microsoft's Database Migration Assistant tool, because your team thinks it doesn't need a DBA and isn't cross-trained on how to properly do a server upgrade and/or didn't bother to read the documentation.  This tool is another easy victory, but a lot of people skip it.  Measure twice, cut once.
   - You're doing a really large change in a single transaction, and the transaction exceeds the amount of transaction log space.  Your test environments don't use databases anywhere near the size of your production tables that the migration touches, and so, while you can even run FluentMigrator in Transaction Per Session mode and SQL Server Transact-SQL will run the migrations, _only you_ can test and/or manually calculate whether your up data migrations will run out of disk space, or create enormous transaction logs, or create VLF fragmentation issues, or whether you configured your system correctly (best practices like setting up instant file initialization).
       - Sometimes, it's best to just use smaller transactions, as they will be more stable.  As an example, converting `ntext` to `nvarchar` requires updating every row in the table, in order to refresh the large object pool. Doing that in a single UPDATE command on a large table is not advisable, but it's possible to split it out into N transactions where N is the number of rows in the table.
   - Somebody wrote a data migration that updated a temporal table's system periods, and the UPDATE statement had a bug that broke the contiguity of the ValidFromInUtc/ValidToInUtc columns.  I usually validate the contiguity automatically and/or peer review `UPDATE` statements that may mess with system periods.
   - FluentMigrator uses one connection for the whole migration process, so **connection-level changes**, that would ordinarily get reset when using connection pooling, **persist across migrations**.  When connection pooling is enabled and you dispose of a connection, System.Data.SqlClient sends an `EXEC sp_reset_connection` command.  This command resets a lot of connection-level details like default database, impersonation context, and any session [SET](https://docs.microsoft.com/en-us/sql/t-sql/statements/set-statements-transact-sql) values. For `TRANSACTION ISOLATION LEVEL`, that is only reset if and only if both of the following are true: (a) you're using SQL Server 2014 or later (b) your client is communicating via TDS 7.3 or higher.  You can check your protocol version via [sys.dm_exec_connections.protocol_version](https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-exec-connections-transact-sql) but it's [not explained](https://denhamcoder.net/2018/08/07/sql-server-sys-dm_exec_connections-protocol_version/) how those values [map to specific SQL Server databases](https://www.freetds.org/userguide/ChoosingTdsProtocol.html).  But, for now, the value is stored in the first byte hex value of the protocol_version: `SELECT SUBSTRING(CAST(protocol_version as binary(4)), 1, 1) FROM sys.dm_exec_connections;`

Long story short: You're using SQL Server.  It supports Transactional DDL.  You're lucky.  Only one other major DBMS also supports Transactional DDL: Postgres.  If you don't have a dedicated DBA, then the magic of Transactional DDL may keep you safe, as long as you're not doing anything complicated that could require an expert to review.  As always, "defense in depth" is a good idea, but the amount of work you put in to "bulletproof" things should probably match the level of your organization.  Not everyone can afford/needs "Fortune 500 controls".  If you're large enough to afford a DBA, hire one.  Consider hiring a consulting DBA to hire that DBA for you, if nobody knows how to screen DBA candidates.  Or just hire a consultant to do part-time DBA work once a week or so.  Not every DBA will like working with FluentMigrator and other schema scaffolding tools that generate SQL DDL.

> 3\. We couldn't find in the docs how to specify a Profile (i.e. Development) that will run certain data migrations on our local dev. environments.

Use Maintenance Migrations, and apply Tags to your Maintenance Migrations, so that only TagName = Development and untagged migrations run.

This is a really long reply, but I do consider myself an expert in SQL Server (and also a C# engineer is my day job).  So, I took the time to explain to you the various gotchas you may not be thinking about if you're not an expert.  I hope you got something out of it.  I do realize better samples would be great.  But... it's also not a request I get very often.  And it's probably best you explain your process rather than just say it's a "CI/CD approach".
